dev/TEAL.md

```diff
@@ -6,62 +6,147 @@ abstract: >
+Smart Contracts have access to everything a Smart Signature may access
+(see previous section), as well as the ability to examine blockchain
+state such as balances and contract state (their own state and the
+state of other contracts).  They also have access to some global
+values that are not visible to Smart Signatures because the values
+change over time.  Since smart contracts access changing state, nodes
+must rerun their code to determine if the ApplicationCall transactions
+in their pool would still succeed each time a block is added to the
+blockchain.
 
-The TEAL bytecode plus the length of all Args must add up to no more than 1000 bytes (consensus parameter LogicSigMaxSize). Each TEAL op has an associated cost and the program cost must total no more than 20000 (consensus parameter LogicSigMaxCost). Most ops have a cost of 1, but a few slow crypto ops are much higher. Prior to v4, the program's cost was estimated as the static sum of all the opcode costs in the program (whether they were actually executed or not). Beginning with v4, the program's cost is tracked dynamically, while being evaluated. If the program exceeds its budget, it fails.
+### Resource availability
+
+Smart contracts have limits on their execution budget (700, consensus
+paramter MaxAppProgramCost), and the amount of blockchain state they
+may examine.  Opcodes may only access blockchain resources such as
+Accounts, Assets, and contract state if the given resource is
+_available_. 
+
+ * A resource in the "foreign array" fields of the ApplicationCall
+   transaction (`txn.Accounts`, `txn.ForeignAssets`, and
+   `txn.ForeignApplications`) is _available_.
+
+ * The `global CurrentApplicationID` and `txn.Sender` are _available_.
+
+ * Prior to v4, all assets were considered _available_ to the
+   `asset_holding_get` opcode.
+
+ * Since v6, any asset or contract that was created earlier in the
+   same transaction group is _available_. In addition, any account
+   that is the contract account of a contract that was created earlier
+   in the group is _available_.
 
 ## Constants
 
-Constants are loaded into the environment into storage separate from the stack. They can then be pushed onto the stack by referring to the type and index. This makes for efficient re-use of byte constants used for account addresses, etc. Constants that are not reused can be pushed with `pushint` or `pushbytes`.
+Constants are loaded into storage separate from the stack and scratch space. They can then be pushed onto the stack by referring to the type and index. This makes for efficient re-use of byte constants used for account addresses, etc. Constants that are not reused can be pushed with `pushint` or `pushbytes`.
```

Comment:
if handled by the compiler, my preference would be to not even mention it here.

Reply:
Yes, I believe the assembler uses `pushint` if the constant is loaded only once.

Reply:
I agree with @jannotti , I think even if hidden by the assembler, this needs to be specified because this is a specs of the low-level language. That being said, no need indeed to indicate specifically what the assembler is supposed to do.

(There is a minor issue not indicating it precisely though: another assembler may yield a different TEAL bytecode from the same TEAL program. But I don't think we need to worry about that right now.)